doc/source/user-profiling.rst

```diff
@@ -0,0 +1,486 @@
+execution time of all function calls made within the profiled code. 
+
+.. _`profiling module`: https://docs.python.org/3/library/profile.html#module-cProfile
+
+Unlike ``line_profiler`` above, this detailed list of profiled function calls 
+**includes** internal function calls and function calls made within Ray! 
+
+However, similar to ``line_profiler``, cProfile can be enabled with minimal 
+changes to your application code (given that each section of the code you want 
+to profile is defined as its own function). To use cProfile, add an import 
+statement, then replace calls to the loop functions as follows:
+
+.. code-block:: python
+
+  import cProfile  # Added import statement
+
+  def ex1():
+    list1 = []
+    for i in range(5):
+      list1.append(ray.get(func.remote()))
+
+  def main():
+    ray.init()
+    cProfile.run('ex1()')  # Modified call to ex1
+    cProfile.run('ex2()')
+    cProfile.run('ex3()')
+
+  if __name__ == "__main__":
+    main()
+
+Now, when executing your Python script, a cProfile list of profiled function 
+calls will be outputted to terminal for each call made to ``cProfile.run()``.
+At the very top of cProfile's output gives the total execution time for 
+``'ex1()'``:
+
+.. code-block:: bash
+
+  601 function calls (595 primitive calls) in 2.509 seconds
+
+Following is a snippet of profiled function calls for ``'ex1()'``. Most of 
+these calls are quick and take around 0.000 seconds, so the functions of 
+interest are the ones with non-zero execution times:
+
+.. code-block:: bash
+
+  ncalls  tottime  percall  cumtime  percall filename:lineno(function)
+  ...
+      1    0.000    0.000    2.509    2.509 your_script_here.py:31(ex1)
+      5    0.000    0.000    0.001    0.000 remote_function.py:103(remote)
+      5    0.000    0.000    0.001    0.000 remote_function.py:107(_submit)
+  ...  
+     10    0.000    0.000    0.000    0.000 worker.py:2459(__init__)
+      5    0.000    0.000    2.508    0.502 worker.py:2535(get)
+      5    0.000    0.000    0.000    0.000 worker.py:2695(get_global_worker)
+     10    0.000    0.000    2.507    0.251 worker.py:374(retrieve_and_deserialize)
+      5    0.000    0.000    2.508    0.502 worker.py:424(get_object)
+      5    0.000    0.000    0.000    0.000 worker.py:514(submit_task)
+  ...
+
+The 5 separate calls to Ray's ``get``, taking the full 0.502 seconds each call, 
+can be noticed at ``worker.py:2535(get)``. Meanwhile, the act of calling the 
+remote function itself at ``remote_function.py:103(remote)`` only takes 0.001 
+seconds over 5 calls, and thus is not the source of the slow performance of 
+``ex1()``.
+
+
+Visualizing Tasks in the Ray Timeline
+-------------------------------------
+Profiling the performance of your Ray application doesn't need to be 
+an eye-straining endeavor of interpreting numbers among hundreds of 
+lines of text. Ray comes with its own visual web UI to visualize the 
+parallelization (or lack thereof) of user tasks submitted to Ray!
```

Comment:
Point out the limitation somewhere, i.e. this only shows timing info about ray tasks, but doesn't break timing down to normal python functions. This can be a problem especially on a driver, that's why we explain the other profiling techniques above.